File: iOS/TRTC-API-Example-OC/ScreenShare/TRTCBroadcastExtensionLauncher.m
//
//  TRTCBroadcastExtensionLauncher.m
//  TRTC-API-Example-OC
//
//  Created by bluedang on 2021/4/15.
//
//


#import <UIKit/UIKit.h>
#import <ReplayKit/ReplayKit.h>

#import "TRTCBroadcastExtensionLauncher.h"

static TRTCBroadcastExtensionLauncher* launch;

API_AVAILABLE(ios(12.0))
@interface TRTCBroadcastExtensionLauncher()
@property (strong, nonatomic) RPSystemBroadcastPickerView* systemBroacastExtensionPicker;
@end

@implementation TRTCBroadcastExtensionLauncher

+ (instancetype) sharedInstance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        launch = [[self alloc] init];
    });
    return launch;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        RPSystemBroadcastPickerView* picker =
                [[RPSystemBroadcastPickerView alloc] initWithFrame:CGRectMake(0, 0, 44, 44)];
        picker.showsMicrophoneButton = false;
        picker.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleRightMargin;
        _systemBroacastExtensionPicker = picker;
        
        NSString *plugInPath = NSBundle.mainBundle.builtInPlugInsPath;
        if (!plugInPath) {
            return self;
        }
        
        NSArray* contents = [NSFileManager.defaultManager contentsOfDirectoryAtPath:plugInPath error:nil];
        for (NSString* content in contents) {
            NSURL* url = [NSURL fileURLWithPath:plugInPath];
            NSBundle* bundle = [NSBundle bundleWithPath:[url URLByAppendingPathComponent:content].path];
            
            NSDictionary* extension = [bundle.infoDictionary objectForKey:@"NSExtension"];
            if (extension == nil) { continue; }
            NSString* identifier = [extension objectForKey:@"NSExtensionPointIdentifier"];
            if ([identifier isEqualToString:@"com.apple.broadcast-services-upload"]) {
                picker.preferredExtension = bundle.bundleIdentifier;
                break;
            }
        }
    }
    return self;
}

+ (void)launch {
    [[TRTCBroadcastExtensionLauncher sharedInstance] launch];
}

- (void)launch {
    for (UIView* view in _systemBroacastExtensionPicker.subviews) {
        UIButton* button = (UIButton*)view;
        [button sendActionsForControlEvents:UIControlEventAllTouchEvents];
    }
}

@end

File: iOS/TRTC-API-Example-Swift/Screen/TRTCBroadcastExtensionLauncher.swift
//
//  TRTCBroadcastExtensionLauncher.swift
//  TRTCSimpleDemo
//
//  Created by J J on 2020/6/10.
//  Copyright © 2020 Tencent. All rights reserved.
//

import UIKit
import ReplayKit

@available(iOS 12.0, *)
class TRTCBroadcastExtensionLauncher: NSObject {
    var systemBroacastExtensionPicker = RPSystemBroadcastPickerView()
    var prevLaunchEventTime : CFTimeInterval = 0
    
    static let sharedInstance =  TRTCBroadcastExtensionLauncher()
    
    override init() {
        super.init()
        let picker = RPSystemBroadcastPickerView(frame: CGRect(x: 0, y: 0, width: 44, height: 44))
        picker.showsMicrophoneButton = false
        picker.autoresizingMask = [.flexibleTopMargin, .flexibleRightMargin]
        systemBroacastExtensionPicker = picker
        
        if let pluginPath = Bundle.main.builtInPlugInsPath,
            let contents = try? FileManager.default.contentsOfDirectory(atPath: pluginPath) {
            
            for content in contents where content.hasSuffix(".appex") {
                guard let bundle = Bundle(path: URL(fileURLWithPath: pluginPath).appendingPathComponent(content).path),
                    let identifier : String = (bundle.infoDictionary?["NSExtension"] as? [String:Any])? ["NSExtensionPointIdentifier"] as? String
                    else {
                        continue
                }
                if identifier == "com.apple.broadcast-services-upload" {
                    picker.preferredExtension = bundle.bundleIdentifier
                    break
                }
            }
        }
    }
    
    static func launch() {
        TRTCBroadcastExtensionLauncher.sharedInstance.launch()
    }
    
    func launch() {
        // iOS 12 上弹出比较慢，如果快速点击会Crash
        let now = CFAbsoluteTimeGetCurrent()
        if now - prevLaunchEventTime < 1.0 {
            return;
        }
        prevLaunchEventTime = now

        for view in systemBroacastExtensionPicker.subviews {
            if let button = view as? UIButton {
                button.sendActions(for: .allTouchEvents)
                break
            }
        }
    }
}
